### PR TITLE
Update core API endpoints

### PR DESCRIPTION
**Pull request by daria305**
Issue #50  

Implement:
 - add client lib functions and tests
 - update staking response with pagination, it contains registered validators not only active
 - rewards endpoint returns epoch range that was used
 - other missing endpoints and fixes
 - implement RMC endpoint

What's left:
 - protocol versioning, probably should be tracked within performance tracker, as we need latest versions of registered validator, not only committee members
 - evalevaluate  event API

### CODE REVIEW

**Review by jkrvivian — CHANGES_REQUESTED** [review]

File: components/restapi/core/accounts.go [hunk 1]
```diff
@@ -31,33 +37,43 @@ func congestionForAccountID(c echo.Context) (*apimodels.CongestionResponse, erro
...
+	pageSize, _ := httpserver.ParseUint32QueryParam(c, restapipkg.QueryParameterPageSize)
+	requestedSlotIndex, cursorIndex, _ := httpserver.ParseCursorQueryParam(c, restapipkg.QueryParameterCursor)
 	latestCommittedSlot := deps.Protocol.SyncManager.LatestCommitment().Index()
-	nextEpoch := deps.Protocol.APIForSlot(latestCommittedSlot).TimeProvider().EpochFromSlot(latestCommittedSlot) + 1
 
-	activeValidators, err := deps.Protocol.MainEngineInstance().SybilProtection.EligibleValidators(nextEpoch)
-	if err != nil {
-		return nil, err
+	if cursorIndex == 0 {
+		requestedSlotIndex = latestCommittedSlot
```
Comment (jkrvivian) [inline components/restapi/core/accounts.go:52]: Why is `requestedSlotIndex` set to `latestCommittedSlot` if `cursorIndex` == 0?
Or how would the request be if requesting the first page of a specific slot index?

Comment (daria305) [reply components/restapi/core/accounts.go]: the cursor index was supposed to be 0 for the first page, that's why I set the requestAtSlotIndex to latestCommittedSlot (that represent current slot and snapshot responses based on it)
But cursor is now string, I could just check for err to see if it was provided. I change also last page response.

File: components/restapi/core/accounts.go [hunk 2]
```diff
@@ -21,49 +28,73 @@ func congestionForAccountID(c echo.Context) (*apimodels.CongestionResponse, erro
...
-	activeValidators, err := deps.Protocol.MainEngineInstance().SybilProtection.EligibleValidators(nextEpoch)
-	if err != nil {
-		return nil, err
+	slotRange := uint32(requestedSlotIndex / RequestsMemoryCacheGranularity)
+	registeredValidators, exists := deps.Protocol.MainEngineInstance().Retainer.RegisteredValidatorsCache(slotRange)
+	if !exists {
+		var err error
+		registeredValidators, err = deps.Protocol.MainEngineInstance().SybilProtection.OrderedRegisteredValidatorsList(nextEpoch)
+		if err != nil {
+			return nil, err
```
Comment (karimodm) [inline components/restapi/core/accounts.go:77]: Wrap the error

File: pkg/protocol/sybilprotection/sybilprotection.go [hunk 3]
```diff
@@ -8,13 +8,16 @@ import (
...
 )
 
 type SybilProtection interface {
 	TrackValidationBlock(block *blocks.Block)
 	EligibleValidators(epoch iotago.EpochIndex) (accounts.AccountsData, error)
-	ValidatorReward(validatorID iotago.AccountID, stakeAmount iotago.BaseToken, epochStart, epochEnd iotago.EpochIndex) (validatorReward iotago.Mana, err error)
-	DelegatorReward(validatorID iotago.AccountID, delegatedAmount iotago.BaseToken, epochStart, epochEnd iotago.EpochIndex) (delegatorsReward iotago.Mana, err error)
+	OrderedRegisteredValidatorsList(epoch iotago.EpochIndex) ([]*apimodels.ValidatorResponse, error)
+	IsActive(validatorID iotago.AccountID, epoch iotago.EpochIndex) bool
+	ValidatorReward(validatorID iotago.AccountID, stakeAmount iotago.BaseToken, epochStart, epochEnd iotago.EpochIndex) (validatorReward iotago.Mana, usedStart, usedEnd iotago.EpochIndex, err error)
```
Comment (karimodm) [inline pkg/protocol/sybilprotection/sybilprotection.go:19]: I think the returned epoch values should have more descriptive names, and you should provide a comment description of what they mean.

File: pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go [hunk 4]
```diff
@@ -311,6 +315,52 @@ func (o *SybilProtection) EligibleValidators(epoch iotago.EpochIndex) (accounts.
 	return validators, nil
 }
 
+// IsActive returns true if the given validator is currently active.
+func (o *SybilProtection) IsActive(validatorID iotago.AccountID, epoch iotago.EpochIndex) bool {
```
Comment (karimodm) [inline pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go:319]: This method should be called IsActiveCandidate, no?

File: pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go [hunk 5]
```diff
@@ -311,6 +315,52 @@ func (o *SybilProtection) EligibleValidators(epoch iotago.EpochIndex) (accounts.
...
+	activeCandidates := o.performanceTracker.EligibleValidatorCandidates(epoch)
+	return activeCandidates.Has(validatorID)
+}
+
+// OrderedRegisteredValidatorsList returns the currently known list of registered validator candidates for the given epoch.
+func (o *SybilProtection) OrderedRegisteredValidatorsList(epoch iotago.EpochIndex) ([]*apimodels.ValidatorResponse, error) {
```
Comment (karimodm) [inline pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go:325]: Should be called OrderedRegisteredCandidateValidatorsList

File: pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go [hunk 6]
```diff
@@ -311,6 +315,52 @@ func (o *SybilProtection) EligibleValidators(epoch iotago.EpochIndex) (accounts.
...
+	validatorResp := make([]*apimodels.ValidatorResponse, 0, candidates.Size())
+	if err := candidates.ForEach(func(candidate iotago.AccountID) error {
+		accountData, exists, err := o.ledger.Account(candidate, o.lastCommittedSlot)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return ierrors.Errorf("account of committee candidate does not exist: %s", candidate)
+		}
+		if accountData.StakeEndEpoch < epoch {
```
Comment (karimodm) [inline pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go:338]: Is the StakeEndEpoch inclusive or exclusive? in other words: is the StakeEndEpoch the last epoch in which your stake is active or is the first epoch such stake doesn't apply anymore?

**Review by karimodm — CHANGES_REQUESTED** [review]

File: pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go (hunk 6 above)
Comment (daria305) [reply pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go]: From TIP: Feature whose `End Epoch` is the current one or has passed, it is no longer considered for validator selection and the. So the cond was incorrect, changed to `<=`

**Review by jkrvivian — APPROVED** [review]

**Review by karimodm — APPROVED** [review]

**Pull request merged**